### ISSUE TITLE
Annotation count is not displayed for offline documents

### ISSUE THREAD

**Issue by tophee** [hypothesis/browser-extension#60]
I you annotate a pdf document in a group (perhaps also for other visibility settings, haven't tested) and open this document online, the hypothesis extension icon will have a blue rectangle with the number of annotations:
![image](https://user-images.githubusercontent.com/3662750/33521417-dbb6b2b4-d7d1-11e7-97c1-b2b4b69b7b14.png)
If you open the same document offline, the annotation count is missing.

**Comment by Pictor13** [comment]
The extension can't query the server when offline. So it won't see any annotation; because it cannot get to know them.

It's intended. I dunno if at Hypothesis are planning to implement any local storage/caching.
But 2 years passed... 🤷‍♂️

**Comment by tophee** [comment]
> The extension can't query the server when offline.

Yes, of course not. I think what I meant was that when you open the file *as a local file* (while, of course still being online), this happens.

**Comment by Pictor13** [comment]
I suppose Hypothesis stores the annotations as assigned to the local/remote resource URL.
If the URL of the online PDF is different from the URL of the downloaded PDF (e.g. `http://github.com/myfile.pdf` for the file when online and `file://users/tophee/downloads/myfile.pdf` for the local file) then Hypothesis sees just two different resources; it won't make any attempt to find out if it's the same file but as local version.
Different URL: different annotated resource.

Inside Hypothesis, when opening on locale, are the annotations appearing instead?
Or are they coherent with the count? (hence they are missing)

**Comment by tophee** [comment]
> I suppose Hypothesis stores the annotations as assigned to the local/remote resource URL.

No, this is not the case. I believe that pdfs are not stored by URL at all but by their internal unique identifier. 

This issue is not about the pdf not being recognised by hypothesis but about the annotation count not being displayed.

**Comment by Pictor13** [comment]
Mmh.. than might be something related to the same issue I have.

I get an incorrect count as well, sometimes like you and sometimes the opposite instead, showing a count but then the page has no annotations.
But in my case it happens also being online.

I don't know if it's a bug. Would be nice to have some feedback from the maintainers.

**Comment by robertknight** [comment]
The basic problem here relates to the URL that Hypothesis uses to identify the document you are looking at. When annotating HTTP URLs we save the URL with the annotation, and then the toolbar can get the same URL from the address bar and query for the annotation count. For file URLs this doesn't work since they are not globally shared (`file:///Users/robert/bar.pdf` on your machine is not the same as that URL on someone else's). Instead Hypothesis relies on a fingerprint embedded in the file, so that everyone sharing a copy of the PDF sees the same annotations, even if the files are stored in different locations on each device. This fingerprint is only available after downloading and parsing the file, and that information is not currently available to the code which fetches the annotation counts in the toolbar.

**Comment by robertknight** [comment]
See also https://github.com/hypothesis/browser-extension/issues/79 which is another issue where querying the badge count with the URL from the address bar, vs content extracted from the document, makes a difference.
